docs: UTXO and VM (#225)
FuelLabs/docs-hub: docs/fuel-101/intro-to-utxo.mdx

---

> title: Intro to UTXO

Reviewer: Intro to the UTXO Model or just UTXO? I think perhaps the former?

Reviewer 2: I think it should be either "Intro to the UTXO Model" or "Intro to UTXOs"

Reviewer: Agreed. What do you prefer?

Author: Ahaha I have a bad habit of misaligned titles. I liked "Intro to the UTXO model" the best. Aligned nav, title, and metadata

category: Fuel 101
---

> # Understanding UTXO

Reviewer: ```suggestion
# Understanding the UTXO Model
```

> The Unspent Transaction Output (UTXO) model, is a fundamental concept in Fuel's underlying blockchain architecture. Fuel tracks digital currency through transaction outputs, each representing a discrete piece of state that can be spent. This model is extended by stateful UTXOs to include additional data or conditions, enabling complex transactions and smart contract functionalities within a UTXO framework.

Reviewer 2: ```suggestion
The Unspent Transaction Output (UTXO) model is a fundamental concept in Fuel's underlying blockchain architecture. Fuel tracks digital currency through transaction outputs, each representing a discrete piece of state that can be spent. This model is extended by stateful UTXOs to include additional data or conditions, enabling complex transactions and smart contract functionalities within a UTXO framework.
```

small nit

> 

Reviewer 2: It feels like there should be a section header here like "How do transactions work with a UTXO model?" or something

> For example, if Bob wants to send Alice three coins, his wallet utilizes a 10-coin UTXO, sends three coins to Alice, and issues seven coins back to Bob as a new UTXO. This process guarantees that the sum of the inputs equals or exceeds the total outputs, a principle that validators can enforce easily. These newly created output UTXOs can then serve as input UTXOs in subsequent transactions.

    <Box.Centered>
    ![UTXO example](/images/utxo.png)
    </Box.Centered>

## Stateful UTXOs

> The UTXO model has often been misunderstood, especially in terms of its ability to support complex functionalities similar to those found in Ethereum style smart contracts. However, Fuel uses an execution model that capitalizes on the strengths of UTXOs while maintaining smart contract capabilities. To enable smart contracts a blockchain need these essential elements.

Reviewer: ```suggestion
The UTXO model has often been misunderstood, especially in terms of its ability to support complex functionalities similar to those found in Ethereum style smart contracts. However, Fuel uses an execution model that capitalizes on the strengths of UTXOs while maintaining smart contract capabilities. To enable smart contracts a blockchain need these essential elements:
```

> The UTXO model has often been misunderstood, especially in terms of its ability to support complex functionalities similar to those found in Ethereum style smart contracts. However, Fuel uses an execution model that capitalizes on the strengths of UTXOs while maintaining smart contract capabilities. To enable smart contracts a blockchain need these essential elements.

Reviewer 2: I think this could be simplified.

```suggestion
Fuel uses an execution model that capitalizes on the strengths of UTXOs while maintaining smart contract capabilities. To enable smart contracts, a blockchain needs these essential elements:
```

### Contract Creation

> Contract UTXOs can be generated through an opcode that deterministically assigns a cryptographically unique contract ID, similar to Ethereum's CREATE2 opcode.

Reviewer 2: nit 

```suggestion
Contract UTXOs can be generated through an opcode that deterministically assigns a cryptographically unique contract ID, similar to Ethereum's `CREATE2` opcode.
```

### Conditional Spending

When a contract UTXO is used in a transaction, it must generate precisely one new UTXO. This new UTXO inherits specific spending conditions, maintaining the original contract's code while updating its state root.

### Cross Contract Calls

By using strict access lists, transactions explicitly list the contracts they intend to interact with, allowing for the seamless interaction among contracts by utilizing their UTXOs in a unified transaction.

> ## Why Choose UTXO Over Account-Based Models?

Reviewer: We should add a section on state minimization as a benefit as well

> The UTXO (Unspent Transaction Output) model offers significant advantages over Ethereum's account-based framework. Fuel's architecture is inherently designed for high scalability, enabling the parallel processing of transactions.

Reviewer 2: ```suggestion
The UTXO (Unspent Transaction Output) model offers significant advantages over Ethereum's account-based framework, including high scalability and enabling the parallel processing of transactions.
```

### Parallelization

> By specifying which contracts are involved in a transaction, it allows transactions that interact with disjoint sets of contracts to be executed concurrently, a feature streamlined through the use of strict access lists. It also enhances transaction validation efficiency by explicitly defining state transitions, permitting parallel validation by non-block-producing nodes, even for transactions that engage the same contracts.

Reviewer 2: ```suggestion
By specifying which contracts are involved in a transaction, the UTXO model allows transactions that interact with disjoint sets of contracts to be executed concurrently, streamlined through the use of strict access lists. It also enhances transaction validation efficiency by explicitly defining state transitions and permitting parallel validation by non-block-producing nodes, even for transactions that engage the same contracts.
```

### Nonce Management

> The UTXO model also simplifies nonce management by using implicit nonces derived from outpoints, thereby eliminating the need to track nonces for addresses with zero balances, an essential aspect of the account-based model.

Reviewer 2: ```suggestion
The UTXO model also simplifies nonce management by using implicit nonces derived from outpoints. This feature eliminates the need to track nonces for addresses with zero balances, an essential aspect of the account-based model.
```
